src/serializer/functions.js

```diff
@@ -122,40 +122,69 @@ export class Functions {
     let recordedReactRootComponents = this.__generateAdditionalFunctions("__reactComponentRoots");
 
     // Get write effects of the components
     for (let [componentType] of recordedReactRootComponents) {
       let simpleClassComponents = new Set();
-      let reconciler = new Reconciler(this.realm, this.moduleTracer, statistics, react, simpleClassComponents);
+      let branchReactComponentTrees = [];
+      let reconciler = new Reconciler(
+        this.realm,
+        this.moduleTracer,
+        statistics,
+        react,
+        simpleClassComponents,
+        branchReactComponentTrees
+      );
       invariant(
         componentType instanceof ECMAScriptSourceFunctionValue,
         "only ECMAScriptSourceFunctionValue function values are supported as React root components"
       );
-      let effects = reconciler.render(componentType, this.moduleTracer.modules.logger);
+      let effects = reconciler.render(componentType, null, null, true);
+      this._generateWriteEffectsForReactComponentTree(componentType, effects, simpleClassComponents);
 
-      if (this.realm.react.output === "bytecode") {
-        throw new FatalError("TODO: implement React bytecode output format");
-      } else {
-        let additionalFunctionEffects = this._createAdditionalEffects(effects);
-        invariant(effects[0] instanceof Value);
-        if (simpleClassComponents.has(effects[0])) {
-          // if the root component was a class and is now simple, we can convert it from a class
-          // component to a functional component
-          convertSimpleClassComponentToFunctionalComponent(this.realm, componentType, additionalFunctionEffects);
-          normalizeFunctionalComponentParamaters(componentType);
-          this.writeEffects.set(componentType, additionalFunctionEffects);
-        } else if (valueIsClassComponent(this.realm, componentType)) {
-          let prototype = Get(this.realm, componentType, "prototype");
-          invariant(prototype instanceof ObjectValue);
-          let renderMethod = Get(this.realm, prototype, "render");
-          invariant(renderMethod instanceof ECMAScriptSourceFunctionValue);
-          this.writeEffects.set(renderMethod, additionalFunctionEffects);
-        } else {
-          normalizeFunctionalComponentParamaters(componentType);
-          this.writeEffects.set(componentType, additionalFunctionEffects);
-        }
+      // for now we just use abstract props/context, in the future we'll create a new branch with a new component
+      // that used the props/context. It will extend the original component and only have a render method
+      for (let { componentType: branchComponentType } of branchReactComponentTrees) {
+        let branchEffects = reconciler.render(branchComponentType, null, null, false);
+        this._generateWriteEffectsForReactComponentTree(branchComponentType, branchEffects, simpleClassComponents);
```

Comment:
Does the order matter here? Can you create a test with multiple trees and different depths?

Reply:
I didn't want to use `Set` here so that in the future we could process the same function multiple times with different `props` and `context`.